dot/network/notifications.go

```diff
@@ -75,57 +75,67 @@ type handshakeReader struct {
 }
 
 type notificationsProtocol struct {
-	protocolID         protocol.ID
-	getHandshake       HandshakeGetter
-	handshakeDecoder   HandshakeDecoder
-	handshakeValidator HandshakeValidator
+	protocolID               protocol.ID
+	getHandshake             HandshakeGetter
+	handshakeDecoder         HandshakeDecoder
+	handshakeValidator       HandshakeValidator
+	outboundHandshakeMutexes *sync.Map //map[peer.ID]*sync.Mutex
+	inboundHandshakeData     *sync.Map //map[peer.ID]*handshakeData
+	outboundHandshakeData    *sync.Map //map[peer.ID]*handshakeData
+}
 
-	inboundHandshakeData  *sync.Map //map[peer.ID]*handshakeData
-	outboundHandshakeData *sync.Map //map[peer.ID]*handshakeData
+func newNotificationsProtocol(protocolID protocol.ID, handshakeGetter HandshakeGetter, handshakeDecoder HandshakeDecoder, handshakeValidator HandshakeValidator) *notificationsProtocol {
+	return &notificationsProtocol{
+		protocolID:               protocolID,
+		getHandshake:             handshakeGetter,
+		handshakeValidator:       handshakeValidator,
+		handshakeDecoder:         handshakeDecoder,
+		outboundHandshakeMutexes: new(sync.Map),
+		inboundHandshakeData:     new(sync.Map),
+		outboundHandshakeData:    new(sync.Map),
+	}
 }
 
-func (n *notificationsProtocol) getInboundHandshakeData(pid peer.ID) (handshakeData, bool) {
+func (n *notificationsProtocol) getInboundHandshakeData(pid peer.ID) (*handshakeData, bool) {
 	var (
 		data interface{}
 		has  bool
 	)
 
 	data, has = n.inboundHandshakeData.Load(pid)
 	if !has {
-		return handshakeData{}, false
+		return nil, false
 	}
 
-	return data.(handshakeData), true
+	return data.(*handshakeData), true
 }
 
-func (n *notificationsProtocol) getOutboundHandshakeData(pid peer.ID) (handshakeData, bool) {
+func (n *notificationsProtocol) getOutboundHandshakeData(pid peer.ID) (*handshakeData, bool) {
 	var (
 		data interface{}
 		has  bool
 	)
 
 	data, has = n.outboundHandshakeData.Load(pid)
 	if !has {
-		return handshakeData{}, false
+		return nil, false
 	}
 
-	return data.(handshakeData), true
+	return data.(*handshakeData), true
```

Comment:
Should we type assert with `handshakeData, ok := data.(*handshakeData)`? How can we be sure it will be the right type?

Reply:
well we are always storing `*handshakeData` in the code, if we aren't, we have definitely coded something wrong and should panic. I can still add the explicit check in there and panic if it isn't the right type if you'd like?

Reply:
No let's keep it as is, it will panic just fine as is :smiley: I was just wondering if it could be something else as well. If not, then let's keep it!